### PR TITLE
Remoteiso

### PR DESCRIPTION
**Pull request by fireballwayne**
Need assistance.  Hrydgard, this used to work before your new mutex routines.  I'm not sure how to fix it.  It hangs when cachefileloader calls mutex.  When it is tries to load the thumbnails.

Changes:
A manual option in the GUI to specify server
Ability to read apache directory index output.
Ability to read from subdirectory on webserver.

### CODE REVIEW

**Comment by hrydgard** [comment]
Thanks for letting me know that it broke, I'll look into it! Probably a recursive_mutex that shouldn't have been changed into a mutex.

**Comment by hrydgard** [comment]
Alright, fixed it. You might want to rebase and test this change on top of the new master?

**Comment by fireballwayne** [comment]
Thanks, It is working now.  It's ready to merge pending your approval.

**Comment by hrydgard** [comment]
Great, I'll read it tomorrow. Looks like a useful feature.

File: UI/RemoteISOScreen.cpp
Comment (hrydgard) [inline UI/RemoteISOScreen.cpp]: Change this to

    if (g_Config.bRemoteISOManual)
          return;

to avoid reindenting the rest of the function.

Also please add a comment explaining why.

File: UI/GameSettingsScreen.cpp
Comment (hrydgard) [inline UI/GameSettingsScreen.cpp]: Better comment please - trying to abide by HTTP rules?

File: UI/RemoteISOScreen.cpp
Comment (hrydgard) [inline UI/RemoteISOScreen.cpp]: Oh. Haha, I forgot I named the server that. Please change both this and the place where it's defined to PPSSPPServer or something.

File: ext/native/base/stringutil.cpp
Comment (hrydgard) [inline ext/native/base/stringutil.cpp]: Rename GetQuotedStrings

File: UI/GameSettingsScreen.cpp
Comment (hrydgard) [inline UI/GameSettingsScreen.cpp]: I wonder if we can think of a better name than "Manual". Hm.

**Review by hrydgard — CHANGES_REQUESTED** [review]

File: UI/GameSettingsScreen.cpp
Comment (fireballwayne) [inline UI/GameSettingsScreen.cpp]: All complete except for this one.  I couldn't think of a better name.  I'm open to suggestions.

File: UI/RemoteISOScreen.cpp
Comment (fireballwayne) [inline UI/RemoteISOScreen.cpp]: Done, but I think we should leave SuperDuperServer in there for now as a check.  So it will work correctly if a v1.3 is the server for a v1.4+ client.

File: UI/GameSettingsScreen.cpp
Comment (unknownbrackets) [inline UI/GameSettingsScreen.cpp]: This seems unnecessarily limiting.  Maybe someone will publish a PHP script you can host on a local server which generates something PPSSPP reads, but uses a query string, like a frontend.

-[Unknown]

File: UI/RemoteISOScreen.cpp
Comment (unknownbrackets) [inline UI/RemoteISOScreen.cpp]: The correct way to handle this would be with a Content-Type (it would be `text/plain` from PPSSPP.)  This would allow other clients to send the same responses without pretending to be PPSSPPServer.

-[Unknown]

**Review by unknownbrackets — CHANGES_REQUESTED** [review]

File: UI/RemoteISOScreen.cpp
Comment (InusualZ) [inline UI/RemoteISOScreen.cpp]: Comment not aligned properly

File: UI/GameSettingsScreen.cpp
Comment (fireballwayne) [inline UI/GameSettingsScreen.cpp]: I agree it should be on the UI for connecting.  I didn't know how to code that with the screen manager, so I put it here.  The remoteisoscreen has 2 columns and I didn't know how to put a linear layout below that spans the entire screen.

File: UI/GameSettingsScreen.cpp
Comment (fireballwayne) [inline UI/GameSettingsScreen.cpp]: I will make this correction.  I just tested + on Apache, it didn't work.

File: UI/GameSettingsScreen.cpp
Comment (fireballwayne) [inline UI/GameSettingsScreen.cpp]: Currently a query string would break the code that creates the game list.  It just combines the subdir with the game filename.  That is why I blocked it.  I would have to strip it off before combining it.  I'll look into it.

File: UI/GameSettingsScreen.cpp
Comment (fireballwayne) [inline UI/GameSettingsScreen.cpp]: This one is a little more difficult.  If the user gives me http://server:80/xyz is xyz a directory or cgi script.  If it is a directory then I have to put a / before the filename, if it is a cgi then I have to strip it off.  I'll have to think about that.

**Comment by fireballwayne** [comment]
Ready for review.

File: UI/RemoteISOScreen.cpp
Comment (unknownbrackets) [inline UI/RemoteISOScreen.cpp]: Definitely not a good idea to assume that `Content-Type` will always be the second header.  Also, should be case insensitive.  I'd think something more like:

```c++
std::string contentType;
for (const std::string &header : responseHeaders) {
    if (startsWithNoCase(header, "Content-Type:")) {
        contentType = header.substr(strlen("Content-Type:"));
        // Strip any whitespace (TODO: maybe move this to stringutil?)
        contentType.erase(0, contentType.find_first_not_of(" \t\r\n"))
        contentType.erase(contentType.find_last_not_of(" \t\r\n") + 1);
    }
}

// TODO: Technically, "TExt/hTml    ; chaRSet    =    Utf8" should pass, but "text/htmlese" should not.
// But unlikely that'll be an issue.
bool parseHtml = startsWithNoCase(contentType, "text/html");
bool parseText = startsWithNoCase(contentType, "text/plain");

if (parseText) {
    // PPSSPP server / plain text response listing of URIs.
} else if (parseHtml) {
    // Parse out URLs.  Warning: some webservers may use relative href values, others absolute.
} else {
    ERROR_LOG(FILESYS, "Unsupported Content-Type: %s", contentType.c_str());
    return false;
}
```

File: UI/RemoteISOScreen.h
Comment (unknownbrackets) [inline UI/RemoteISOScreen.h]: Just delete this line, no need to have it commented out.

-[Unknown]

**Review by unknownbrackets — CHANGES_REQUESTED** [review]
Sorry, just want to make sure we don't back into a situation where an innocuous change to the HTTP server breaks the header detection.

-[Unknown]

File: UI/RemoteISOScreen.cpp
Comment (fireballwayne) [inline UI/RemoteISOScreen.cpp]: Thanks for the assist.

**Review by unknownbrackets — APPROVED** [review]

**Comment by hrydgard** [comment]
Good catch unknown.

I'm happy with this now too, so I'll merge.

**Pull request merged**
